File: src/llm/azure-openai-embedding.ts
import OpenAI, { AzureOpenAI } from 'openai';
import { withBackoff } from '../pipeline/concurrency.js';
import { CliError } from '../utils/errors.js';
import { batchEmbed, validateProbeDimensions } from './embed-batch.js';
import type { EmbeddingProvider } from './embedding-provider.js';

const BATCH = 96;
const DEFAULT_MODEL = 'text-embedding-3-small';
const DEFAULT_DIMENSIONS = 1536;
const DEFAULT_API_VERSION = '2024-06-01';

export interface AzureOpenaiEmbeddingOptions {
  apiKey: string;
  endpoint: string;
  apiVersion?: string;
  deployment?: string;
  model?: string;
  /**
   * Vector width. Override for non-1536 deployments (e.g.
   * `text-embedding-3-large` = 3072) or to request reduced output via
   * the API's `dimensions` parameter.
   */
  dimensions?: number;
}

export class AzureOpenaiEmbeddingProvider implements EmbeddingProvider {
  readonly name = 'azure-openai';
  readonly dimensions: number;
  private client: AzureOpenAI;
  private model: string;
  private explicitDimensions: number | undefined;

  constructor(opts: AzureOpenaiEmbeddingOptions) {
    this.client = new AzureOpenAI({
      apiKey: opts.apiKey,
      endpoint: opts.endpoint,
      apiVersion: opts.apiVersion ?? DEFAULT_API_VERSION,
      deployment: opts.deployment,
    });
    this.model = opts.model ?? DEFAULT_MODEL;
    this.dimensions = opts.dimensions ?? DEFAULT_DIMENSIONS;
    this.explicitDimensions = opts.dimensions;
  }

  private buildParams(
    input: string | string[]
  ): OpenAI.Embeddings.EmbeddingCreateParams {
    const params: OpenAI.Embeddings.EmbeddingCreateParams = {
      model: this.model,
      input,
    };
    if (this.explicitDimensions !== undefined) {
      params.dimensions = this.explicitDimensions;
    }
    return params;
  }

  async probe(): Promise<void> {
    let response: OpenAI.Embeddings.CreateEmbeddingResponse;
    try {
      response = await this.client.embeddings.create(this.buildParams('probe'));
    } catch (err) {
      throw wrapAzureError('Azure OpenAI embedding probe failed', err);
    }
    validateProbeDimensions({
      actual: response.data?.[0]?.embedding.length,
      declared: this.dimensions,
      providerLabel: 'Azure OpenAI',
      model: this.model,
      remediationHint: `Set GDRIVESCOPE_AZURE_OPENAI_EMBEDDING_DIMENSIONS=${response.data?.[0]?.embedding.length} or pick a different deployment.`,
    });
  }

  embed(texts: string[]): Promise<number[][]> {
    return batchEmbed(texts, BATCH, async (batch) => {
      try {
        const response = await withBackoff(() =>
          this.client.embeddings.create(this.buildParams([...batch]))
        );
        return response.data.map((item) => item.embedding);
      } catch (err) {
        throw wrapAzureError('Azure OpenAI embedding call failed', err);
      }
    });
  }
}

function wrapAzureError(prefix: string, err: unknown): CliError {
  if (err instanceof OpenAI.APIError) {
    return new CliError(
      `${prefix} (status ${err.status}): ${err.message}`,
      'EMBED_CALL_FAILED'
    );
  }
  const msg = err instanceof Error ? err.message : String(err);
  return new CliError(`${prefix}: ${msg}`, 'EMBED_CALL_FAILED');
}

File: src/llm/azure_openai_embedding.rs
//! Azure OpenAI embeddings. Ported from `src/llm/azure-openai-embedding.ts`.
//! URL shape:
//!   {endpoint}/openai/deployments/{deployment}/embeddings?api-version=...

use async_trait::async_trait;
use serde::Deserialize;
use serde_json::json;

use super::embed_batch::{batch_embed, validate_probe_dimensions, ProbeOptions};
use super::embedding::EmbeddingProvider;
use crate::error::{CliError, ErrorCode};

const BATCH: usize = 96;
const DEFAULT_MODEL: &str = "text-embedding-3-small";
const DEFAULT_DIMENSIONS: usize = 1536;
const DEFAULT_API_VERSION: &str = "2024-06-01";

#[derive(Debug, Deserialize)]
struct EmbeddingsResponse {
    data: Vec<EmbeddingItem>,
}

#[derive(Debug, Deserialize)]
struct EmbeddingItem {
    embedding: Vec<f32>,
}

pub struct AzureOpenaiEmbeddingProvider {
    http: reqwest::Client,
    api_key: String,
    endpoint: String,
    api_version: String,
    deployment: String,
    model: String,
    dimensions: usize,
    explicit_dimensions: Option<usize>,
}

pub struct AzureOpenaiEmbeddingOptions {
    pub api_key: String,
    pub endpoint: String,
    pub api_version: Option<String>,
    pub deployment: Option<String>,
    pub model: Option<String>,
    pub dimensions: Option<usize>,
}

impl AzureOpenaiEmbeddingProvider {
    pub fn new(opts: AzureOpenaiEmbeddingOptions) -> Self {
        let model = opts.model.unwrap_or_else(|| DEFAULT_MODEL.to_string());
        let deployment = opts.deployment.unwrap_or_else(|| model.clone());
        Self {
            http: reqwest::Client::new(),
            api_key: opts.api_key,
            endpoint: opts.endpoint.trim_end_matches('/').to_string(),
            api_version: opts
                .api_version
                .unwrap_or_else(|| DEFAULT_API_VERSION.to_string()),
            deployment,
            model,
            dimensions: opts.dimensions.unwrap_or(DEFAULT_DIMENSIONS),
            explicit_dimensions: opts.dimensions,
        }
    }

    fn url(&self) -> String {
        format!(
            "{}/openai/deployments/{}/embeddings?api-version={}",
            self.endpoint, self.deployment, self.api_version,
        )
    }

    fn build_body(&self, input: &serde_json::Value) -> serde_json::Value {
        let mut body = json!({ "model": &self.model, "input": input });
        if let Some(d) = self.explicit_dimensions {
            body["dimensions"] = json!(d);
        }
        body
    }

    async fn call(&self, body: &serde_json::Value) -> Result<EmbeddingsResponse, CliError> {
        let resp = self
            .http
            .post(self.url())
            .header("api-key", &self.api_key)
            .json(body)
            .send()
            .await
            .map_err(|e| {
                CliError::new(
                    format!("Azure OpenAI embedding call failed: {e}"),
                    ErrorCode::EmbedCallFailed,
                )
            })?;
        if !resp.status().is_success() {
            let status = resp.status();
            let text = resp.text().await.unwrap_or_default();
            return Err(CliError::new(
                format!("Azure OpenAI embedding call failed (status {status}): {text}"),
                ErrorCode::EmbedCallFailed,
            ));
        }
        resp.json().await.map_err(|e| {
            CliError::new(
                format!("Azure OpenAI embedding parse failed: {e}"),
                ErrorCode::EmbedCallFailed,
            )
        })
    }
}

#[async_trait]
impl EmbeddingProvider for AzureOpenaiEmbeddingProvider {
    fn name(&self) -> &str {
        "azure-openai"
    }
    fn dimensions(&self) -> usize {
        self.dimensions
    }

    async fn embed(&self, texts: &[String]) -> Result<Vec<Vec<f32>>, CliError> {
        batch_embed(texts, BATCH, |batch| async move {
            let body = self.build_body(&json!(batch));
            let resp = self.call(&body).await?;
            Ok(resp.data.into_iter().map(|d| d.embedding).collect())
        })
        .await
    }

    async fn probe(&self) -> Result<(), CliError> {
        let body = self.build_body(&json!("probe"));
        let resp = self.call(&body).await?;
        validate_probe_dimensions(ProbeOptions {
            actual: resp.data.first().map(|d| d.embedding.len()),
            declared: self.dimensions,
            provider_label: "Azure OpenAI",
            model: &self.model,
            remediation_hint:
                "Set GDRIVESCOPE_AZURE_OPENAI_EMBEDDING_DIMENSIONS or pick a different deployment.",
        })
    }
}
